File: in_class_ex/Comparison Sort Exercise/c++/merge_sort.cpp
#include <iostream>
#include <vector>

// Merge two subarrays of arr[]
// First subarray is arr[l..m]
// Second subarray is arr[m+1..r]
void merge(std::vector<int>& arr, int l, int m, int r) {
    int n1 = m - l + 1;
    int n2 = r - m;

    // Create temporary arrays
    std::vector<int> L(n1);
    std::vector<int> R(n2);

    // Put your code here

}

// Main function to perform Merge Sort
void mergeSort(std::vector<int>& arr, int l, int r) {
    if (l < r) {
        // Same as (l+r)/2, but avoids overflow for large l and r
        int m = l + (r - l) / 2;

        // Sort first and second halves
        // Put your code here

        // Merge the sorted halves
        merge(arr, l, m, r);
    }
}

// Function to print an array
void printArray(const std::vector<int>& arr) {
    for (int num : arr) {
        std::cout << num << " ";
    }
    std::cout << std::endl;
}

int main() {
    std::vector<int> arr = {12, 11, 13, 5, 6, 7};
    int n = arr.size();

    std::cout << "Original Array: ";
    printArray(arr);

    mergeSort(arr, 0, n - 1);

    std::cout << "Sorted Array: ";
    printArray(arr);

    return 0;
}

File: in_class_ex/Comparison Sort Exercise/python/merge_sort.py
def merge(arr, left, right):
    i = j = k = 0

    # Put your code here


def merge_sort(arr):
    if len(arr) > 1:
        mid = len(arr) // 2
        left_half = arr[:mid]
        right_half = arr[mid:]

        # Put your code here

        merge(arr, left_half, right_half)

# Example usage:
arr = [12, 11, 13, 5, 6, 7]
print("Original Array:", arr)
merge_sort(arr)
print("Sorted Array:", arr)
